File: python/structures/src/stack/priority_queue.py
# ordered linked list implementation of a priority queue
class PriorityQueue(object):
    def __init__(self):
        self._values = None

    def __max__(self):
        return self._values.value if self._values else None

    def enqueue(self, value, key):
        to_insert = _PriorityQueueNode(value, key)

        if self._values == None:
            self._values = to_insert
        elif key > self._values.key:
            to_insert.next = self._values
            self._values = to_insert
        else:
            curr = self._values
            while curr.next != None and curr.next.key >= key:
                curr = curr.next

            nxt = curr.next
            curr.next = to_insert
            curr.next.next = nxt


    def extract_max(self):
        if self._values:
            retval = self._values.value
            self._values = self._values.next if self._values.next else None
            return retval
        else:
            return None

    def __len__(self):
        return len(self._values) if self._values != None else 0


class _PriorityQueueNode(object):
    def __init__(self, value, key):
        self.key = key
        self.value = value
        self.next = None

    def __len__(self):
        return 1 if self.next == None else len(self.next) + 1

File: ruby/src/stack/priority_queue.rb
class PriorityQueue
    def initialize()
        @data = nil
    end

    def enqueue(val, key)
        to_insert = PriorityQueueNode.new(val, key)

        if @data == nil
            @data = to_insert
        elsif key > @data.key
            to_insert.next = @data
            @data = to_insert
        else
            curr = @data

            while curr.next != nil && curr.next.key >= key
                curr = curr.next
            end

            if @curr.next != nil
                to_insert.next = @curr.next.next
            end

            curr.next = to_insert
        end
    end

    def extract_max
        if @data == nil
            raise RangeError, 'attempted to extract from empty queue'
        end

        retval = @data.value
        @data = @data.next
        retval
    end

    def size
        if @data
            @data.length()
        else
            0
        end
    end
end

class PriorityQueueNode
    attr_reader :value, :key, :next
    attr_writer :value, :key, :next

    def initialize(val, key)
        @value = val
        @key = key
        @next = nil
    end

    def length
        if @next != nil
            @next.length() + 1
        else
            1
        end
    end
end
